File: test/Zorglub.Tests/Tests/Hemerology/Scopes/BoundedBelowScopeTests.fs
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2020 Narvalo.Org. All rights reserved.

module Zorglub.Tests.Hemerology.Scopes.BoundedBelowScopeTests

open Zorglub.Testing

open Zorglub.Time
open Zorglub.Time.Core.Schemas
open Zorglub.Time.Hemerology.Scopes

open Xunit

module Factories =
    [<Fact>]
    let ``Create() throws for null schema`` () =
        nullExn "schema" (fun () -> BoundedBelowScope.Create(null, DayZero.NewStyle, new DateParts(1, 1, 1), 2))

    [<Fact>]
    let ``Create() throws for an invalid date`` () =
        let sch = new GregorianSchema()

        outOfRangeExn "parts" (fun () -> BoundedBelowScope.Create(sch, DayZero.NewStyle, new DateParts(GregorianSchema.MinYear - 1, 12, 1), GregorianSchema.MaxYear))
        outOfRangeExn "parts" (fun () -> BoundedBelowScope.Create(sch, DayZero.NewStyle, new DateParts(GregorianSchema.MaxYear + 1, 12, 1), GregorianSchema.MaxYear))
        outOfRangeExn "parts" (fun () -> BoundedBelowScope.Create(sch, DayZero.NewStyle, new DateParts(1, 0, 1), 2))
        outOfRangeExn "parts" (fun () -> BoundedBelowScope.Create(sch, DayZero.NewStyle, new DateParts(1, 13, 1), 2))
        outOfRangeExn "parts" (fun () -> BoundedBelowScope.Create(sch, DayZero.NewStyle, new DateParts(1, 1, 0), 2))
        outOfRangeExn "parts" (fun () -> BoundedBelowScope.Create(sch, DayZero.NewStyle, new DateParts(1, 1, 32), 2))

    [<Fact>]
    let ``Create() throws for an invalid max year`` () =
        outOfRangeExn "year" (fun () -> BoundedBelowScope.Create(new GregorianSchema(), DayZero.NewStyle, new DateParts(1, 12, 1), GregorianSchema.MaxYear + 1))

    [<Fact>]
    let ``Create()`` () =
        let parts = new DateParts(5, 3, 13)
        let maxYear = 15
        let scope = BoundedBelowScope.Create(new GregorianSchema(), DayZero.NewStyle, parts, maxYear)
        let seg = scope.Segment

        scope.IsComplete |> nok
        seg.IsComplete |> nok

        seg.MinMaxDateParts.LowerValue === parts
        seg.MinMaxDateParts.UpperValue === new DateParts(maxYear, 12, 31)
        seg.SupportedYears.Max === maxYear

    [<Fact>]
    let ``Create() throws when the min date is the start of a year`` () =
        let parts = new DateParts(5, 1, 1)
        let maxYear = 15

        argExn "segment" (fun () -> BoundedBelowScope.Create(new GregorianSchema(), DayZero.NewStyle, parts, maxYear))

    [<Fact>]
    let ``StartingAt()`` () =
        let scope = BoundedBelowScope.StartingAt(new GregorianSchema(), DayZero.NewStyle, new DateParts(1, 12, 1))
        let seg = scope.Segment

        scope.IsComplete |> nok
        seg.IsComplete |> nok

        seg.SupportedYears.Max === GregorianSchema.MaxYear

File: test/Zorglub.Tests.CSharp/Time/Hemerology/BoundedBelowScopeTests.cs
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2020 Narvalo.Org. All rights reserved.

namespace Zorglub.Time.Hemerology;

using Zorglub.Time.Core;
using Zorglub.Time.Core.Schemas;
using Zorglub.Time.Hemerology.Scopes;

// TODO: on ne vérifie pas le nom du paramètre de certaines exceptions car ce
// dernier varie pas mal: Yemoda.Of(), SystemSegmentBuilder, etc.

public sealed partial class BoundedBelowScopeTests : CalendricalDataConsumer<GregorianDataSet>
{
    private const int FirstYear = 3;
    private const int FirstMonth = 4;
    private const int FirstDay = 5;

    private static readonly ICalendricalSchema s_Schema = new GregorianSchema();

    [Fact]
    public static void Create_NullSchema() =>
        Assert.ThrowsAnexn("schema",
            () => BoundedBelowScope.Create(null!, DayZero.NewStyle, new(3, 4, 5), 9999));

    [Fact]
    public static void Create_InvalidMinYear() =>
        Assert.Throws<ArgumentOutOfRangeException>(
            () => BoundedBelowScope.Create(
                s_Schema, DayZero.NewStyle, new(s_Schema.SupportedYears.Min - 1, 1, 1), 9999));

    [Fact]
    public static void Create_InvalidMaxYear() =>
        Assert.Throws<ArgumentOutOfRangeException>(
            () => BoundedBelowScope.Create(
                s_Schema, DayZero.NewStyle, new(s_Schema.SupportedYears.Max + 1, 1, 1), 9999));

    [Theory, MemberData(nameof(InvalidMonthFieldData))]
    public static void Create_InvalidMonth(int y, int m) =>
        Assert.Throws<ArgumentOutOfRangeException>(
            () => BoundedBelowScope.Create(s_Schema, DayZero.NewStyle, new(y, m, 1), 9999));

    [Theory, MemberData(nameof(InvalidDayFieldData))]
    public static void Create_InvalidDay(int y, int m, int d) =>
        Assert.Throws<ArgumentOutOfRangeException>(
            () => BoundedBelowScope.Create(s_Schema, DayZero.NewStyle, new(y, m, d), 9999));

    [Theory, MemberData(nameof(DateInfoData))]
    public void Create(DateInfo info)
    {
        var (y, m, d, doy) = info;
        // Act
        var scope = BoundedBelowScope.Create(s_Schema, DayZero.NewStyle, new(y, m, d), 9999);
        var minDate = scope.Segment.MinMaxDateParts.LowerValue;
        var minOrdinalDate = scope.Segment.MinMaxOrdinalParts.LowerValue;
        // Assert
        Assert.NotNull(scope);
        Assert.Equal(y, scope.YearsValidator.MinYear);
        Assert.Equal(y, minDate.Year);
        Assert.Equal(m, minDate.Month);
        Assert.Equal(d, minDate.Day);
        Assert.Equal(9999, scope.YearsValidator.MaxYear);
        Assert.Equal(doy, minOrdinalDate.DayOfYear);
    }

    [Fact]
    public static void ValidateYearMonth()
    {
        var scope = BoundedBelowScope.Create(
            s_Schema, DayZero.NewStyle, new(FirstYear, FirstMonth, FirstDay), 9999);
        // Act
        scope.ValidateYearMonth(FirstYear, FirstMonth);
        Assert.ThrowsAoorexn("month", () => scope.ValidateYearMonth(FirstYear, FirstMonth - 1));
    }

    [Fact]
    public static void ValidateYearMonthDay()
    {
        var scope = BoundedBelowScope.Create(
            s_Schema, DayZero.NewStyle, new(FirstYear, FirstMonth, FirstDay), 9999);
        // Act
        scope.ValidateYearMonthDay(FirstYear, FirstMonth, FirstDay);
        scope.ValidateYearMonthDay(FirstYear, FirstMonth, FirstDay + 1);
        scope.ValidateYearMonthDay(FirstYear, FirstMonth + 1, 1);
        Assert.ThrowsAoorexn("month", () => scope.ValidateYearMonthDay(FirstYear, FirstMonth - 1, FirstDay));
        Assert.ThrowsAoorexn("day", () => scope.ValidateYearMonthDay(FirstYear, FirstMonth, FirstDay - 1));
    }

    [Fact]
    public static void ValidateOrdinal()
    {
        int firstDayOfYear = s_Schema.CountDaysInYearBeforeMonth(FirstYear, FirstMonth) + FirstDay;
        var scope = BoundedBelowScope.Create(
            s_Schema, DayZero.NewStyle, new(FirstYear, FirstMonth, FirstDay), 9999);
        // Act
        scope.ValidateOrdinal(FirstYear, firstDayOfYear);
        Assert.ThrowsAoorexn("dayOfYear", () => scope.ValidateOrdinal(FirstYear, firstDayOfYear - 1));
    }
}
